Add Nvidia Triton Server for GPU Deployment

yuekaizhang opened the pull request:
GPU Depolyment Support with Nvidia [Triton](https://github.com/triton-inference-server/tensorrtllm_backend) and [TensorRT-LLM](https://github.com/NVIDIA/TensorRT-LLM)

Launch:
`cd runtime/triton_trtllm && docker compose up`

Performance:
On a single L20 GPU, test with https://huggingface.co/datasets/yuekai/seed_tts_cosy2
| Model | Note   | Concurrency | Avg Latency     | RTF | 
|-------|-----------|-----------------------|---------|--|
| Spark-TTS-0.5B |Triton + TensorRT-LLM | 2                   | 920.97 ms | 0.0737|

weedge commented:
`pip show omegaconf`:
```
Name: omegaconf
Version: 2.3.0
Summary: A flexible configuration library
Home-page: https://github.com/omry/omegaconf
Author: Omry Yadan
Author-email: omry@yadan.net
License: 
Location: /usr/local/lib/python3.10/site-packages
Requires: antlr4-python3-runtime, PyYAML
Required-by: 
```
run `tritonserver --model-repository {model_repo}` bash run.sh 3 3  have some error:
```
I0313 15:24:04.021326 262 pb_stub.cc:320]  Failed to initialize Python stub for auto-complete: ModuleNotFoundError: No module named 'omegaconf'

At:
  /Spark-TTS/sparktts/models/bicodec.py(20): <module>
  <frozen importlib._bootstrap>(488): _call_with_frames_removed
  <frozen importlib._bootstrap_external>(995): exec_module
  <frozen importlib._bootstrap>(950): _load_unlocked
  <frozen importlib._bootstrap>(1334): _find_and_load_unlocked
  <frozen importlib._bootstrap>(1360): _find_and_load
  /root/tritonserver/tts-spark/vocoder/1/model.py(37): <module>
  <frozen importlib._bootstrap>(488): _call_with_frames_removed
  <frozen importlib._bootstrap_external>(995): exec_module
  <frozen importlib._bootstrap>(950): _load_unlocked
  <frozen importlib._bootstrap>(1334): _find_and_load_unlocked
  <frozen importlib._bootstrap>(1360): _find_and_load

E0313 15:24:04.040097 168 model_lifecycle.cc:654] "failed to load 'vocoder' version 1: Internal: ModuleNotFoundError: No module named 'omegaconf'\n\nAt:\n  /Spark-TTS/sparktts/models/bicodec.py(20): <module>\n  <frozen importlib._bootstrap>(488): _call_with_frames_removed\n  <frozen importlib._bootstrap_external>(995): exec_module\n  <frozen importlib._bootstrap>(950): _load_unlocked\n  <frozen importlib._bootstrap>(1334): _find_and_load_unlocked\n  <frozen importlib._bootstrap>(1360): _find_and_load\n  /root/tritonserver/tts-spark/vocoder/1/model.py(37): <module>\n  <frozen importlib._bootstrap>(488): _call_with_frames_removed\n  <frozen importlib._bootstrap_external>(995): exec_module\n  <frozen importlib._bootstrap>(950): _load_unlocked\n  <frozen importlib._bootstrap>(1334): _find_and_load_unlocked\n  <frozen importlib._bootstrap>(1360): _find_and_load\n"
I0313 15:24:04.040204 168 model_lifecycle.cc:789] "failed to load 'vocoder'"
I0313 15:24:11.761819 209 pb_stub.cc:320]  Failed to initialize Python stub for auto-complete: ModuleNotFoundError: No module named 'omegaconf'
```
@yuekaizhang  Could you please help me see what the problem is？

yuekaizhang commented:
> `pip show omegaconf`:
> 
> ```
> Name: omegaconf
> Version: 2.3.0
> Summary: A flexible configuration library
> Home-page: https://github.com/omry/omegaconf
> Author: Omry Yadan
> Author-email: omry@yadan.net
> License: 
> Location: /usr/local/lib/python3.10/site-packages
> Requires: antlr4-python3-runtime, PyYAML
> Required-by: 
> ```
> 
> run `tritonserver --model-repository {model_repo}` bash run.sh 3 3 have some error:
> 
> ```
> I0313 15:24:04.021326 262 pb_stub.cc:320]  Failed to initialize Python stub for auto-complete: ModuleNotFoundError: No module named 'omegaconf'
> 
> At:
>   /Spark-TTS/sparktts/models/bicodec.py(20): <module>
>   <frozen importlib._bootstrap>(488): _call_with_frames_removed
>   <frozen importlib._bootstrap_external>(995): exec_module
>   <frozen importlib._bootstrap>(950): _load_unlocked
>   <frozen importlib._bootstrap>(1334): _find_and_load_unlocked
>   <frozen importlib._bootstrap>(1360): _find_and_load
>   /root/tritonserver/tts-spark/vocoder/1/model.py(37): <module>
>   <frozen importlib._bootstrap>(488): _call_with_frames_removed
>   <frozen importlib._bootstrap_external>(995): exec_module
>   <frozen importlib._bootstrap>(950): _load_unlocked
>   <frozen importlib._bootstrap>(1334): _find_and_load_unlocked
>   <frozen importlib._bootstrap>(1360): _find_and_load
> 
> E0313 15:24:04.040097 168 model_lifecycle.cc:654] "failed to load 'vocoder' version 1: Internal: ModuleNotFoundError: No module named 'omegaconf'\n\nAt:\n  /Spark-TTS/sparktts/models/bicodec.py(20): <module>\n  <frozen importlib._bootstrap>(488): _call_with_frames_removed\n  <frozen importlib._bootstrap_external>(995): exec_module\n  <frozen importlib._bootstrap>(950): _load_unlocked\n  <frozen importlib._bootstrap>(1334): _find_and_load_unlocked\n  <frozen importlib._bootstrap>(1360): _find_and_load\n  /root/tritonserver/tts-spark/vocoder/1/model.py(37): <module>\n  <frozen importlib._bootstrap>(488): _call_with_frames_removed\n  <frozen importlib._bootstrap_external>(995): exec_module\n  <frozen importlib._bootstrap>(950): _load_unlocked\n  <frozen importlib._bootstrap>(1334): _find_and_load_unlocked\n  <frozen importlib._bootstrap>(1360): _find_and_load\n"
> I0313 15:24:04.040204 168 model_lifecycle.cc:789] "failed to load 'vocoder'"
> I0313 15:24:11.761819 209 pb_stub.cc:320]  Failed to initialize Python stub for auto-complete: ModuleNotFoundError: No module named 'omegaconf'
> ```
> 
> @yuekaizhang Could you please help me see what the problem is？

Are you using the prebuilt image? soar97/triton-spark-tts:25.02
Also, you could try to force reinstall the lib.

weedge commented:
noop~  use modal to deploy,  run tritonserver have error....

- download hf ckpt
```shell
modal run download_models.py --repo-ids "SparkAudio/Spark-TTS-0.5B"
```

download_models.py
```python
import modal

app = modal.App("download_models")

# We also define the dependencies for our Function by specifying an
# [Image](https://modal.com/docs/guide/images).

download_image = (
    modal.Image.debian_slim(python_version="3.11")
    .apt_install()
    .run_commands()
    .pip_install("hf-transfer", "huggingface_hub")
    .env(
        {"HF_HUB_ENABLE_HF_TRANSFER": "1"},  # hf-transfer for faster downloads
    )
)

HF_MODEL_DIR = "/root/models"
hf_model_vol = modal.Volume.from_name("models", create_if_missing=True)

# ASSETS_DIR = "/root/assets"
# assets_dir = modal.Volume.from_name("assets", create_if_missing=True)


@app.function(
    # gpu="T4",
    retries=0,
    cpu=8.0,
    image=download_image,
    volumes={HF_MODEL_DIR: hf_model_vol},
    timeout=1200,
    scaledown_window=1200,
)
def download_ckpt(repo_ids: str) -> str:
    import os

    # https://huggingface.co/docs/huggingface_hub/guides/download
    from huggingface_hub import snapshot_download

    for repo_id in repo_ids.split(","):
        local_dir = os.path.join(HF_MODEL_DIR, repo_id)
        snapshot_download(
            repo_id=repo_id,
            repo_type="model",
            allow_patterns="*",
            # ignore_patterns=["*.pt", "*.bin"],  # using safetensors
            local_dir=local_dir,
        )
        print(f"{repo_id} model to dir:{HF_MODEL_DIR} done")


"""
modal run src/download_models.py --repo-ids "stepfun-ai/Step-Audio-Chat,stepfun-ai/Step-Audio-TTS-3B,stepfun-ai/Step-Audio-Tokenizer"
modal run src/download_models.py --repo-ids "SparkAudio/Spark-TTS-0.5B"
"""


@app.local_entrypoint()
def main(repo_ids: str):
    download_ckpt.remote(repo_ids)

```

- compile to build trt-llm engine
```shell
modal run  compile_model.py \
    --app-name "tts-spark" \
    --hf-repo-dir "SparkAudio/Spark-TTS-0.5B/LLM" \
    --trt-dtype "bfloat16" \
    --convert-script-url "https://raw.githubusercontent.com/SparkAudio/Spark-TTS/refs/heads/main/runtime/triton_trtllm/scripts/convert_checkpoint.py" \
    --convert-other-args "" \
    --compile-other-args "--max_batch_size 16 --max_num_tokens 32768"
```

compile_model.py 
```python
import os
import modal

app = modal.App("trtllm")

trtllm_image = (
    # https://catalog.ngc.nvidia.com/orgs/nvidia/containers/cuda/tags
    modal.Image.from_registry(
        "nvidia/cuda:12.8.0-devel-ubuntu22.04",
        add_python="3.10",
    )
    .entrypoint([])  # remove verbose logging by base image on entry
    .apt_install(
        "git", "git-lfs", "openmpi-bin", "libopenmpi-dev", "wget"
    )  # OpenMPI for distributed communication
    .pip_install(
        "tensorrt-llm==0.17.0.post1",
        "omegaconf==2.3.0",
        # "pynvml<12",  # avoid breaking change to pynvml version API for tensorrt_llm
        # "tensorrt==10.8.0.43",
        pre=True,
        extra_index_url="https://pypi.nvidia.com",
    )
    .env({})
)


HF_MODEL_DIR = "/root/models"
hf_model_vol = modal.Volume.from_name("models", create_if_missing=True)
TRT_MODEL_DIR = "/root/trt_models"
trt_model_vol = modal.Volume.from_name("triton_trtllm_models", create_if_missing=True)


GIT_TAG_OR_HASH = "v0.16.0"
CONVERSION_SCRIPT_URL = f"https://raw.githubusercontent.com/NVIDIA/TensorRT-LLM/{GIT_TAG_OR_HASH}/examples/qwen/convert_checkpoint.py"


@app.function(
    gpu=os.getenv("IMAGE_GPU", "L4"),
    retries=0,
    image=trtllm_image,
    volumes={
        HF_MODEL_DIR: hf_model_vol,
        TRT_MODEL_DIR: trt_model_vol,
    },
    timeout=1200,  # default 300s
    scaledown_window=1200,
)
def trtllm_build(
    app_name: str,
    hf_repo_dir: str,
    trt_dtype: str = "bfloat16",
    convert_script_url: str = CONVERSION_SCRIPT_URL,
    convert_other_args: str = "",
    compile_other_args: str = "",
) -> str:
    import subprocess

    cmd = f"pip show tensorrt".split(" ")
    subprocess.run(cmd, cwd="/", check=True)

    cmd = f"which nvcc".split(" ")
    subprocess.run(cmd, cwd="/", check=True)
    cmd = f"nvcc --version".split(" ")
    subprocess.run(cmd, cwd="/", check=True)

    cmd = f"ls {HF_MODEL_DIR}/{hf_repo_dir}".split(" ")
    subprocess.run(cmd, cwd="/")

    cmd = f"wget {convert_script_url} -O /root/convert.py".split(" ")
    subprocess.run(cmd, cwd="/")

    local_hf_model_dir = os.path.join(HF_MODEL_DIR, hf_repo_dir)
    local_trt_model_dir = os.path.join(TRT_MODEL_DIR, app_name, f"tllm_checkpoint_{trt_dtype}")

    print("「Convert」 Converting checkpoint to TensorRT weights")
    cmd = (
        f"python /root/convert.py --model_dir {local_hf_model_dir} "
        + f"--output_dir {local_trt_model_dir} --dtype {trt_dtype} "
        + f"{convert_other_args}"
    )
    print(cmd)
    subprocess.run(cmd.strip().split(" "), cwd="/")

    print("「Compilation」Building TensorRT engines")
    local_trt_build_dir = os.path.join(TRT_MODEL_DIR, app_name, f"trt_engines_{trt_dtype}")
    cmd = (
        f"trtllm-build --checkpoint_dir {local_trt_model_dir} --output_dir {local_trt_build_dir} "
        + f"--gemm_plugin {trt_dtype} "
        + f"{compile_other_args}"
    )
    print(cmd)
    subprocess.run(cmd.strip().split(" "), cwd="/")


"""
modal run src/llm/trtllm/compile_model.py \
    --app-name "tts-spark" \
    --hf-repo-dir "SparkAudio/Spark-TTS-0.5B/LLM" \
    --trt-dtype "bfloat16" \
    --convert-other-args "" \
    --compile-other-args "--max_batch_size 16 --max_num_tokens 32768"

modal run src/llm/trtllm/compile_model.py \
    --app-name "tts-spark" \
    --hf-repo-dir "SparkAudio/Spark-TTS-0.5B/LLM" \
    --trt-dtype "bfloat16" \
    --convert-script-url "https://raw.githubusercontent.com/SparkAudio/Spark-TTS/refs/heads/main/runtime/triton_trtllm/scripts/convert_checkpoint.py" \
    --convert-other-args "" \
    --compile-other-args "--max_batch_size 16 --max_num_tokens 32768"
"""


@app.local_entrypoint()
def main(
    app_name: str,
    hf_repo_dir: str,
    trt_dtype: str = "bfloat16",
    convert_script_url: str = CONVERSION_SCRIPT_URL,
    convert_other_args: str = "",
    compile_other_args: str = "",
):
    trtllm_build.remote(
        app_name,
        hf_repo_dir,
        trt_dtype,
        convert_script_url,
        convert_other_args,
        compile_other_args,
    )

```

- ready tritonserver model api params 
```shell
# fill template with pbtext file for api params
modal run ready_model.py \
    --tag-or-hash "main" \
    --trt-dtype "bfloat16" \
    --spark-tts-params "bls_instance_num:4,triton_max_batch_size:16,max_queue_delay_microseconds:0" \
    --audio-tokenizer-params "triton_max_batch_size:16,max_queue_delay_microseconds:0" \
    --tensorrt-llm-params "triton_backend:tensorrtllm,triton_max_batch_size:16,decoupled_mode:False,max_beam_width:1,max_tokens_in_paged_kv_cache:2560,max_attention_window_size:2560,kv_cache_free_gpu_mem_fraction:0.5,exclude_input_in_output:True,enable_kv_cache_reuse:False,batching_strategy:inflight_fused_batching,max_queue_delay_microseconds:0,encoder_input_features_data_type:TYPE_FP16,logits_datatype:TYPE_FP32" \
    --vocoder-params "triton_max_batch_size:16,max_queue_delay_microseconds:0" 
```
ready_model.py
```python
import os
import modal

app_name = "tts-spark"
app = modal.App(f"{app_name}-tritonserver")

# Define the dependencies for the function using a Modal Image.
image = modal.Image.debian_slim(python_version="3.10").apt_install("git")
image = image.env(
    {
        "HF_REPO_ID": os.getenv("HF_REPO_ID", "SparkAudio/Spark-TTS-0.5B"),
    }
)

HF_MODEL_DIR = "/root/models"
hf_model_vol = modal.Volume.from_name("models", create_if_missing=True)
TRT_MODEL_DIR = "/root/trt_models"
trt_model_vol = modal.Volume.from_name("triton_trtllm_models", create_if_missing=True)
TRITONSERVER_DIR = "/root/tritonserver"
tritonserver_vol = modal.Volume.from_name("tritonserver", create_if_missing=True)


# see: https://github.com/triton-inference-server/python_backend/blob/main/README.md
@app.function(
    retries=0,
    image=image,
    volumes={
        HF_MODEL_DIR: hf_model_vol,
        TRT_MODEL_DIR: trt_model_vol,
        TRITONSERVER_DIR: tritonserver_vol,
    },
    timeout=1200,  # default 300s
    scaledown_window=1200,
)
def ready_model(
    spark_tts_params: str,
    audio_tokenizer_params: str,
    tensorrt_llm_params: str,
    vocoder_params: str,
    trt_dtype: str,
    tag_or_hash: str,
) -> str:
    import subprocess

    if tag_or_hash == "":
        tag_or_hash = "main"

    cmd = f"git clone https://github.com/SparkAudio/Spark-TTS -b {tag_or_hash}".split(" ")
    subprocess.run(cmd, cwd="/")

    model_repo = os.path.join(TRITONSERVER_DIR, app_name)
    cmd = f"rm -rf {model_repo}".split(" ")
    subprocess.run(cmd, cwd="/")

    cmd = f"cp -r /Spark-TTS/runtime/triton_trtllm/model_repo {model_repo}".split(" ")
    subprocess.run(cmd, cwd="/")

    hf_model_local_dir = os.path.join(
        HF_MODEL_DIR, os.getenv("HF_REPO_ID", "SparkAudio/Spark-TTS-0.5B")
    )

    # spark tts (Combine 「 audio_tokenizer -> trt llm enginer -> vocoder 」 services)
    hf_model_local_llm_dir = os.path.join(hf_model_local_dir, "LLM")
    cmd = f"python3 scripts/fill_template.py -i {model_repo}/spark_tts/config.pbtxt" + (
        f" {spark_tts_params},llm_tokenizer_dir:{hf_model_local_llm_dir}".strip(",")
    )
    print(cmd)
    subprocess.run(cmd.split(" "), cwd="/Spark-TTS/runtime/triton_trtllm/")

    # audio_tokenizer
    cmd = f"python3 scripts/fill_template.py -i {model_repo}/audio_tokenizer/config.pbtxt" + (
        f" {audio_tokenizer_params},model_dir:{hf_model_local_dir}".strip(",")
    )
    print(cmd)
    subprocess.run(cmd.split(" "), cwd="/Spark-TTS/runtime/triton_trtllm/")

    # trt llm enginer
    local_trt_build_dir = os.path.join(TRT_MODEL_DIR, app_name, f"trt_engines_{trt_dtype}")
    cmd = f"python3 scripts/fill_template.py -i {model_repo}/tensorrt_llm/config.pbtxt" + (
        f" {tensorrt_llm_params},engine_dir:{local_trt_build_dir }".strip(",")
    )
    print(cmd)
    subprocess.run(cmd.split(" "), cwd="/Spark-TTS/runtime/triton_trtllm/")

    # vocoder
    cmd = f"python3 scripts/fill_template.py -i {model_repo}/vocoder/config.pbtxt" + (
        f" {vocoder_params},model_dir:{hf_model_local_dir}".strip(",")
    )
    print(cmd)
    subprocess.run(cmd.split(" "), cwd="/Spark-TTS/runtime/triton_trtllm/")


"""
# fill template with pbtext file for api params
modal run src/llm/trtllm/bench/tts_spark/ready_model.py \
    --tag-or-hash "main" \
    --trt-dtype "bfloat16" \
    --spark-tts-params "bls_instance_num:4,triton_max_batch_size:16,max_queue_delay_microseconds:0" \
    --audio-tokenizer-params "triton_max_batch_size:16,max_queue_delay_microseconds:0" \
    --tensorrt-llm-params "triton_backend:tensorrtllm,triton_max_batch_size:16,decoupled_mode:False,max_beam_width:1,max_tokens_in_paged_kv_cache:2560,max_attention_window_size:2560,kv_cache_free_gpu_mem_fraction:0.5,exclude_input_in_output:True,enable_kv_cache_reuse:False,batching_strategy:inflight_fused_batching,max_queue_delay_microseconds:0,encoder_input_features_data_type:TYPE_FP16,logits_datatype:TYPE_FP32" \
    --vocoder-params "triton_max_batch_size:16,max_queue_delay_microseconds:0" 
"""


@app.local_entrypoint()
def main(
    spark_tts_params: str,
    audio_tokenizer_params: str,
    tensorrt_llm_params: str,
    vocoder_params: str,
    trt_dtype: str = "bfloat16",
    tag_or_hash: str = "main",
):
    ready_model.remote(
        spark_tts_params,
        audio_tokenizer_params,
        tensorrt_llm_params,
        vocoder_params,
        trt_dtype,
        tag_or_hash,
    )

```

- run tritonserver
```shell
modal run tritonserver.py --app-name tts-spark
```

tritonserver.py 
```python
import os
import modal

app = modal.App("tritonserver")

tritonserver_image = (
    # https://catalog.ngc.nvidia.com/orgs/nvidia/containers/tritonserver
    modal.Image.from_registry(
        "nvcr.io/nvidia/tritonserver:25.02-trtllm-python-py3",
        add_python="3.10",
    )
    .entrypoint([])  # remove verbose logging by base image on entry
    .apt_install("git", "git-lfs", "cmake")
    .pip_install(
        #    "einx==0.3.0",
        #    "numpy",
        #    "omegaconf==2.3.0",
        #    "soundfile==0.12.1",
        #    "soxr==0.5.0.post1",
        #    "transformers==4.46.2",
        #    "librosa",
        #    # extra_index_url="https://pypi.nvidia.com",
    )
    .run_commands(
        "git clone https://github.com/SparkAudio/Spark-TTS.git",
        "cd Spark-TTS && pip install -q -r requirements.txt",
        # "git clone https://github.com/pytorch/audio.git",
        # "cd audio && git checkout c670ad8 && PATH=/usr/local/cuda/bin:$PATH python3 setup.py develop",
    )
    .env(
        {
            "PYTHONPATH": os.getenv("IMAGE_PYTHONPATH", "/Spark-TTS/"),
            "CUDA_VISIBLE_DEVICES": os.getenv("CUDA_VISIBLE_DEVICES", "0"),
            "PYTHONIOENCODING": "utf-8",
        }
    )
)


HF_MODEL_DIR = "/root/models"
hf_model_vol = modal.Volume.from_name("models", create_if_missing=True)
TRT_MODEL_DIR = "/root/trt_models"
trt_model_vol = modal.Volume.from_name("triton_trtllm_models", create_if_missing=True)
TRITONSERVER_DIR = "/root/tritonserver"
tritonserver_vol = modal.Volume.from_name("tritonserver", create_if_missing=True)


# ⭐️ https://github.com/triton-inference-server/python_backend/blob/main/README.md
@app.function(
    gpu=os.getenv("IMAGE_GPU", "L4"),
    image=tritonserver_image,
    volumes={
        HF_MODEL_DIR: hf_model_vol,
        TRT_MODEL_DIR: trt_model_vol,
        TRITONSERVER_DIR: tritonserver_vol,
    },
    timeout=1200,  # default 300s
    scaledown_window=1200,
)
# @modal.web_server(port=8000, startup_timeout=5 * 60)
def serve(app_name: str) -> str:
    import subprocess

    os.environ["PYTHONPATH"] = f"/usr/local/lib/python3.10/site-packages:{os.environ['PYTHONPATH']}"

    cmd = f"ldd /opt/tritonserver/bin/tritonserver".split(" ")
    subprocess.run(cmd, cwd="/")
    # cmd = f"ldd /opt/tritonserver/backends/tensorrtllm/libtriton_tensorrtllm.so".split(" ")
    # subprocess.run(cmd, cwd="/")

    model_repo = os.path.join(TRITONSERVER_DIR, app_name)

    cmd = f"tritonserver --model-repository {model_repo}"
    print(cmd)
    subprocess.run(cmd, env=os.environ, shell=True)


"""
# run tritonserver
modal run src/llm/trtllm/bench/tritonserver.py --app-name tts-spark
"""


@app.local_entrypoint()
def main(app_name: str = ""):
    serve.remote(app_name)

```
run cmd `tritonserver --model-repository {model_repo}` have some errors:

```
I0313 17:55:29.117742 80 pb_stub.cc:320]  Failed to initialize Python stub for auto-complete: ImportError: Error importing numpy: you should not try to import numpy from
        its source directory; please exit the numpy source tree, and relaunch
        your python interpreter from there.

At:
  /usr/local/lib/python3.10/site-packages/numpy/__init__.py(119): <module>
  <frozen importlib._bootstrap>(488): _call_with_frames_removed
  <frozen importlib._bootstrap_external>(995): exec_module
  <frozen importlib._bootstrap>(950): _load_unlocked
  <frozen importlib._bootstrap>(1333): _find_and_load_unlocked
  <frozen importlib._bootstrap>(1360): _find_and_load
  /root/tritonserver/tts-spark/spark_tts/1/model.py(32): <module>
  <frozen importlib._bootstrap>(488): _call_with_frames_removed
  <frozen importlib._bootstrap_external>(995): exec_module
  <frozen importlib._bootstrap>(950): _load_unlocked
  <frozen importlib._bootstrap>(1334): _find_and_load_unlocked
  <frozen importlib._bootstrap>(1360): _find_and_load

E0313 17:55:29.121481 25 model_lifecycle.cc:654] "failed to load 'spark_tts' version 1: Internal: ImportError: Error importing numpy: you should not try to import numpy from\n        its source directory; please exit the numpy source tree, and relaunch\n        your python interpreter from there.\n\nAt:\n  /usr/local/lib/python3.10/site-packages/numpy/__init__.py(119): <module>\n  <frozen importlib._bootstrap>(488): _call_with_frames_removed\n  <frozen importlib._bootstrap_external>(995): exec_module\n  <frozen importlib._bootstrap>(950): _load_unlocked\n  <frozen importlib._bootstrap>(1333): _find_and_load_unlocked\n  <frozen importlib._bootstrap>(1360): _find_and_load\n  /root/tritonserver/tts-spark/spark_tts/1/model.py(32): <module>\n  <frozen importlib._bootstrap>(488): _call_with_frames_removed\n  <frozen importlib._bootstrap_external>(995): exec_module\n  <frozen importlib._bootstrap>(950): _load_unlocked\n  <frozen importlib._bootstrap>(1334): _find_and_load_unlocked\n  <frozen importlib._bootstrap>(1360): _find_and_load\n"
I0313 17:55:29.121766 25 model_lifecycle.cc:789] "failed to load 'spark_tts'"
I0313 17:55:29.259034 81 pb_stub.cc:320]  Failed to initialize Python stub for auto-complete: ImportError: Failed to load PyTorch C extensions:
    It appears that PyTorch has loaded the `torch/_C` folder
    of the PyTorch repository rather than the C extensions which
    are expected in the `torch._C` namespace. This can occur when
    using the `install` workflow. e.g.
        $ python setup.py install && python -c "import torch"

    This error can generally be solved using the `develop` workflow
        $ python setup.py develop && python -c "import torch"  # This should succeed
    or by running Python from a different directory.

At:
  /usr/local/lib/python3.10/site-packages/torch/__init__.py(898): <module>
  <frozen importlib._bootstrap>(488): _call_with_frames_removed
  <frozen importlib._bootstrap_external>(995): exec_module
  <frozen importlib._bootstrap>(950): _load_unlocked
  <frozen importlib._bootstrap>(1333): _find_and_load_unlocked
  <frozen importlib._bootstrap>(1360): _find_and_load
  /root/tritonserver/tts-spark/audio_tokenizer/1/model.py(27): <module>
  <frozen importlib._bootstrap>(488): _call_with_frames_removed
  <frozen importlib._bootstrap_external>(995): exec_module
  <frozen importlib._bootstrap>(950): _load_unlocked
  <frozen importlib._bootstrap>(1334): _find_and_load_unlocked
  <frozen importlib._bootstrap>(1360): _find_and_load

E0313 17:55:29.262042 25 model_lifecycle.cc:654] "failed to load 'audio_tokenizer' version 1: Internal: ImportError: Failed to load PyTorch C extensions:\n    It appears that PyTorch has loaded the `torch/_C` folder\n    of the PyTorch repository rather than the C extensions which\n    are expected in the `torch._C` namespace. This can occur when\n    using the `install` workflow. e.g.\n        $ python setup.py install && python -c \"import torch\"\n\n    This error can generally be solved using the `develop` workflow\n        $ python setup.py develop && python -c \"import torch\"  # This should succeed\n    or by running Python from a different directory.\n\nAt:\n  /usr/local/lib/python3.10/site-packages/torch/__init__.py(898): <module>\n  <frozen importlib._bootstrap>(488): _call_with_frames_removed\n  <frozen importlib._bootstrap_external>(995): exec_module\n  <frozen importlib._bootstrap>(950): _load_unlocked\n  <frozen importlib._bootstrap>(1333): _find_and_load_unlocked\n  <frozen importlib._bootstrap>(1360): _find_and_load\n  /root/tritonserver/tts-spark/audio_tokenizer/1/model.py(27): <module>\n  <frozen importlib._bootstrap>(488): _call_with_frames_removed\n  <frozen importlib._bootstrap_external>(995): exec_module\n  <frozen importlib._bootstrap>(950): _load_unlocked\n  <frozen importlib._bootstrap>(1334): _find_and_load_unlocked\n  <frozen importlib._bootstrap>(1360): _find_and_load\n"
```

yuekaizhang commented:
> noop~ use modal to deploy, run tritonserver have error....
> 
> * download hf ckpt
> 
> ```shell
> modal run download_models.py --repo-ids "SparkAudio/Spark-TTS-0.5B"
> ```
> 
> download_models.py
> 
> ```python
> import modal
> 
> app = modal.App("download_models")
> 
> # We also define the dependencies for our Function by specifying an
> # [Image](https://modal.com/docs/guide/images).
> 
> download_image = (
>     modal.Image.debian_slim(python_version="3.11")
>     .apt_install()
>     .run_commands()
>     .pip_install("hf-transfer", "huggingface_hub")
>     .env(
>         {"HF_HUB_ENABLE_HF_TRANSFER": "1"},  # hf-transfer for faster downloads
>     )
> )
> 
> HF_MODEL_DIR = "/root/models"
> hf_model_vol = modal.Volume.from_name("models", create_if_missing=True)
> 
> # ASSETS_DIR = "/root/assets"
> # assets_dir = modal.Volume.from_name("assets", create_if_missing=True)
> 
> 
> @app.function(
>     # gpu="T4",
>     retries=0,
>     cpu=8.0,
>     image=download_image,
>     volumes={HF_MODEL_DIR: hf_model_vol},
>     timeout=1200,
>     scaledown_window=1200,
> )
> def download_ckpt(repo_ids: str) -> str:
>     import os
> 
>     # https://huggingface.co/docs/huggingface_hub/guides/download
>     from huggingface_hub import snapshot_download
> 
>     for repo_id in repo_ids.split(","):
>         local_dir = os.path.join(HF_MODEL_DIR, repo_id)
>         snapshot_download(
>             repo_id=repo_id,
>             repo_type="model",
>             allow_patterns="*",
>             # ignore_patterns=["*.pt", "*.bin"],  # using safetensors
>             local_dir=local_dir,
>         )
>         print(f"{repo_id} model to dir:{HF_MODEL_DIR} done")
> 
> 
> """
> modal run src/download_models.py --repo-ids "stepfun-ai/Step-Audio-Chat,stepfun-ai/Step-Audio-TTS-3B,stepfun-ai/Step-Audio-Tokenizer"
> modal run src/download_models.py --repo-ids "SparkAudio/Spark-TTS-0.5B"
> """
> 
> 
> @app.local_entrypoint()
> def main(repo_ids: str):
>     download_ckpt.remote(repo_ids)
> ```
> 
> * compile to build trt-llm engine
> 
> ```shell
> modal run  compile_model.py \
>     --app-name "tts-spark" \
>     --hf-repo-dir "SparkAudio/Spark-TTS-0.5B/LLM" \
>     --trt-dtype "bfloat16" \
>     --convert-script-url "https://raw.githubusercontent.com/SparkAudio/Spark-TTS/refs/heads/main/runtime/triton_trtllm/scripts/convert_checkpoint.py" \
>     --convert-other-args "" \
>     --compile-other-args "--max_batch_size 16 --max_num_tokens 32768"
> ```
> 
> compile_model.py
> 
> ```python
> import os
> import modal
> 
> app = modal.App("trtllm")
> 
> trtllm_image = (
>     # https://catalog.ngc.nvidia.com/orgs/nvidia/containers/cuda/tags
>     modal.Image.from_registry(
>         "nvidia/cuda:12.8.0-devel-ubuntu22.04",
>         add_python="3.10",
>     )
>     .entrypoint([])  # remove verbose logging by base image on entry
>     .apt_install(
>         "git", "git-lfs", "openmpi-bin", "libopenmpi-dev", "wget"
>     )  # OpenMPI for distributed communication
>     .pip_install(
>         "tensorrt-llm==0.17.0.post1",
>         "omegaconf==2.3.0",
>         # "pynvml<12",  # avoid breaking change to pynvml version API for tensorrt_llm
>         # "tensorrt==10.8.0.43",
>         pre=True,
>         extra_index_url="https://pypi.nvidia.com",
>     )
>     .env({})
> )
> 
> 
> HF_MODEL_DIR = "/root/models"
> hf_model_vol = modal.Volume.from_name("models", create_if_missing=True)
> TRT_MODEL_DIR = "/root/trt_models"
> trt_model_vol = modal.Volume.from_name("triton_trtllm_models", create_if_missing=True)
> 
> 
> GIT_TAG_OR_HASH = "v0.16.0"
> CONVERSION_SCRIPT_URL = f"https://raw.githubusercontent.com/NVIDIA/TensorRT-LLM/{GIT_TAG_OR_HASH}/examples/qwen/convert_checkpoint.py"
> 
> 
> @app.function(
>     gpu=os.getenv("IMAGE_GPU", "L4"),
>     retries=0,
>     image=trtllm_image,
>     volumes={
>         HF_MODEL_DIR: hf_model_vol,
>         TRT_MODEL_DIR: trt_model_vol,
>     },
>     timeout=1200,  # default 300s
>     scaledown_window=1200,
> )
> def trtllm_build(
>     app_name: str,
>     hf_repo_dir: str,
>     trt_dtype: str = "bfloat16",
>     convert_script_url: str = CONVERSION_SCRIPT_URL,
>     convert_other_args: str = "",
>     compile_other_args: str = "",
> ) -> str:
>     import subprocess
> 
>     cmd = f"pip show tensorrt".split(" ")
>     subprocess.run(cmd, cwd="/", check=True)
> 
>     cmd = f"which nvcc".split(" ")
>     subprocess.run(cmd, cwd="/", check=True)
>     cmd = f"nvcc --version".split(" ")
>     subprocess.run(cmd, cwd="/", check=True)
> 
>     cmd = f"ls {HF_MODEL_DIR}/{hf_repo_dir}".split(" ")
>     subprocess.run(cmd, cwd="/")
> 
>     cmd = f"wget {convert_script_url} -O /root/convert.py".split(" ")
>     subprocess.run(cmd, cwd="/")
> 
>     local_hf_model_dir = os.path.join(HF_MODEL_DIR, hf_repo_dir)
>     local_trt_model_dir = os.path.join(TRT_MODEL_DIR, app_name, f"tllm_checkpoint_{trt_dtype}")
> 
>     print("「Convert」 Converting checkpoint to TensorRT weights")
>     cmd = (
>         f"python /root/convert.py --model_dir {local_hf_model_dir} "
>         + f"--output_dir {local_trt_model_dir} --dtype {trt_dtype} "
>         + f"{convert_other_args}"
>     )
>     print(cmd)
>     subprocess.run(cmd.strip().split(" "), cwd="/")
> 
>     print("「Compilation」Building TensorRT engines")
>     local_trt_build_dir = os.path.join(TRT_MODEL_DIR, app_name, f"trt_engines_{trt_dtype}")
>     cmd = (
>         f"trtllm-build --checkpoint_dir {local_trt_model_dir} --output_dir {local_trt_build_dir} "
>         + f"--gemm_plugin {trt_dtype} "
>         + f"{compile_other_args}"
>     )
>     print(cmd)
>     subprocess.run(cmd.strip().split(" "), cwd="/")
> 
> 
> """
> modal run src/llm/trtllm/compile_model.py \
>     --app-name "tts-spark" \
>     --hf-repo-dir "SparkAudio/Spark-TTS-0.5B/LLM" \
>     --trt-dtype "bfloat16" \
>     --convert-other-args "" \
>     --compile-other-args "--max_batch_size 16 --max_num_tokens 32768"
> 
> modal run src/llm/trtllm/compile_model.py \
>     --app-name "tts-spark" \
>     --hf-repo-dir "SparkAudio/Spark-TTS-0.5B/LLM" \
>     --trt-dtype "bfloat16" \
>     --convert-script-url "https://raw.githubusercontent.com/SparkAudio/Spark-TTS/refs/heads/main/runtime/triton_trtllm/scripts/convert_checkpoint.py" \
>     --convert-other-args "" \
>     --compile-other-args "--max_batch_size 16 --max_num_tokens 32768"
> """
> 
> 
> @app.local_entrypoint()
> def main(
>     app_name: str,
>     hf_repo_dir: str,
>     trt_dtype: str = "bfloat16",
>     convert_script_url: str = CONVERSION_SCRIPT_URL,
>     convert_other_args: str = "",
>     compile_other_args: str = "",
> ):
>     trtllm_build.remote(
>         app_name,
>         hf_repo_dir,
>         trt_dtype,
>         convert_script_url,
>         convert_other_args,
>         compile_other_args,
>     )
> ```
> 
> * ready tritonserver model api params
> 
> ```shell
> # fill template with pbtext file for api params
> modal run ready_model.py \
>     --tag-or-hash "main" \
>     --trt-dtype "bfloat16" \
>     --spark-tts-params "bls_instance_num:4,triton_max_batch_size:16,max_queue_delay_microseconds:0" \
>     --audio-tokenizer-params "triton_max_batch_size:16,max_queue_delay_microseconds:0" \
>     --tensorrt-llm-params "triton_backend:tensorrtllm,triton_max_batch_size:16,decoupled_mode:False,max_beam_width:1,max_tokens_in_paged_kv_cache:2560,max_attention_window_size:2560,kv_cache_free_gpu_mem_fraction:0.5,exclude_input_in_output:True,enable_kv_cache_reuse:False,batching_strategy:inflight_fused_batching,max_queue_delay_microseconds:0,encoder_input_features_data_type:TYPE_FP16,logits_datatype:TYPE_FP32" \
>     --vocoder-params "triton_max_batch_size:16,max_queue_delay_microseconds:0" 
> ```
> 
> ready_model.py
> 
> ```python
> import os
> import modal
> 
> app_name = "tts-spark"
> app = modal.App(f"{app_name}-tritonserver")
> 
> # Define the dependencies for the function using a Modal Image.
> image = modal.Image.debian_slim(python_version="3.10").apt_install("git")
> image = image.env(
>     {
>         "HF_REPO_ID": os.getenv("HF_REPO_ID", "SparkAudio/Spark-TTS-0.5B"),
>     }
> )
> 
> HF_MODEL_DIR = "/root/models"
> hf_model_vol = modal.Volume.from_name("models", create_if_missing=True)
> TRT_MODEL_DIR = "/root/trt_models"
> trt_model_vol = modal.Volume.from_name("triton_trtllm_models", create_if_missing=True)
> TRITONSERVER_DIR = "/root/tritonserver"
> tritonserver_vol = modal.Volume.from_name("tritonserver", create_if_missing=True)
> 
> 
> # see: https://github.com/triton-inference-server/python_backend/blob/main/README.md
> @app.function(
>     retries=0,
>     image=image,
>     volumes={
>         HF_MODEL_DIR: hf_model_vol,
>         TRT_MODEL_DIR: trt_model_vol,
>         TRITONSERVER_DIR: tritonserver_vol,
>     },
>     timeout=1200,  # default 300s
>     scaledown_window=1200,
> )
> def ready_model(
>     spark_tts_params: str,
>     audio_tokenizer_params: str,
>     tensorrt_llm_params: str,
>     vocoder_params: str,
>     trt_dtype: str,
>     tag_or_hash: str,
> ) -> str:
>     import subprocess
> 
>     if tag_or_hash == "":
>         tag_or_hash = "main"
> 
>     cmd = f"git clone https://github.com/SparkAudio/Spark-TTS -b {tag_or_hash}".split(" ")
>     subprocess.run(cmd, cwd="/")
> 
>     model_repo = os.path.join(TRITONSERVER_DIR, app_name)
>     cmd = f"rm -rf {model_repo}".split(" ")
>     subprocess.run(cmd, cwd="/")
> 
>     cmd = f"cp -r /Spark-TTS/runtime/triton_trtllm/model_repo {model_repo}".split(" ")
>     subprocess.run(cmd, cwd="/")
> 
>     hf_model_local_dir = os.path.join(
>         HF_MODEL_DIR, os.getenv("HF_REPO_ID", "SparkAudio/Spark-TTS-0.5B")
>     )
> 
>     # spark tts (Combine 「 audio_tokenizer -> trt llm enginer -> vocoder 」 services)
>     hf_model_local_llm_dir = os.path.join(hf_model_local_dir, "LLM")
>     cmd = f"python3 scripts/fill_template.py -i {model_repo}/spark_tts/config.pbtxt" + (
>         f" {spark_tts_params},llm_tokenizer_dir:{hf_model_local_llm_dir}".strip(",")
>     )
>     print(cmd)
>     subprocess.run(cmd.split(" "), cwd="/Spark-TTS/runtime/triton_trtllm/")
> 
>     # audio_tokenizer
>     cmd = f"python3 scripts/fill_template.py -i {model_repo}/audio_tokenizer/config.pbtxt" + (
>         f" {audio_tokenizer_params},model_dir:{hf_model_local_dir}".strip(",")
>     )
>     print(cmd)
>     subprocess.run(cmd.split(" "), cwd="/Spark-TTS/runtime/triton_trtllm/")
> 
>     # trt llm enginer
>     local_trt_build_dir = os.path.join(TRT_MODEL_DIR, app_name, f"trt_engines_{trt_dtype}")
>     cmd = f"python3 scripts/fill_template.py -i {model_repo}/tensorrt_llm/config.pbtxt" + (
>         f" {tensorrt_llm_params},engine_dir:{local_trt_build_dir }".strip(",")
>     )
>     print(cmd)
>     subprocess.run(cmd.split(" "), cwd="/Spark-TTS/runtime/triton_trtllm/")
> 
>     # vocoder
>     cmd = f"python3 scripts/fill_template.py -i {model_repo}/vocoder/config.pbtxt" + (
>         f" {vocoder_params},model_dir:{hf_model_local_dir}".strip(",")
>     )
>     print(cmd)
>     subprocess.run(cmd.split(" "), cwd="/Spark-TTS/runtime/triton_trtllm/")
> 
> 
> """
> # fill template with pbtext file for api params
> modal run src/llm/trtllm/bench/tts_spark/ready_model.py \
>     --tag-or-hash "main" \
>     --trt-dtype "bfloat16" \
>     --spark-tts-params "bls_instance_num:4,triton_max_batch_size:16,max_queue_delay_microseconds:0" \
>     --audio-tokenizer-params "triton_max_batch_size:16,max_queue_delay_microseconds:0" \
>     --tensorrt-llm-params "triton_backend:tensorrtllm,triton_max_batch_size:16,decoupled_mode:False,max_beam_width:1,max_tokens_in_paged_kv_cache:2560,max_attention_window_size:2560,kv_cache_free_gpu_mem_fraction:0.5,exclude_input_in_output:True,enable_kv_cache_reuse:False,batching_strategy:inflight_fused_batching,max_queue_delay_microseconds:0,encoder_input_features_data_type:TYPE_FP16,logits_datatype:TYPE_FP32" \
>     --vocoder-params "triton_max_batch_size:16,max_queue_delay_microseconds:0" 
> """
> 
> 
> @app.local_entrypoint()
> def main(
>     spark_tts_params: str,
>     audio_tokenizer_params: str,
>     tensorrt_llm_params: str,
>     vocoder_params: str,
>     trt_dtype: str = "bfloat16",
>     tag_or_hash: str = "main",
> ):
>     ready_model.remote(
>         spark_tts_params,
>         audio_tokenizer_params,
>         tensorrt_llm_params,
>         vocoder_params,
>         trt_dtype,
>         tag_or_hash,
>     )
> ```
> 
> * run tritonserver
> 
> ```shell
> modal run tritonserver.py --app-name tts-spark
> ```
> 
> tritonserver.py
> 
> ```python
> import os
> import modal
> 
> app = modal.App("tritonserver")
> 
> tritonserver_image = (
>     # https://catalog.ngc.nvidia.com/orgs/nvidia/containers/tritonserver
>     modal.Image.from_registry(
>         "nvcr.io/nvidia/tritonserver:25.02-trtllm-python-py3",
>         add_python="3.10",
>     )
>     .entrypoint([])  # remove verbose logging by base image on entry
>     .apt_install("git", "git-lfs", "cmake")
>     .pip_install(
>         #    "einx==0.3.0",
>         #    "numpy",
>         #    "omegaconf==2.3.0",
>         #    "soundfile==0.12.1",
>         #    "soxr==0.5.0.post1",
>         #    "transformers==4.46.2",
>         #    "librosa",
>         #    # extra_index_url="https://pypi.nvidia.com",
>     )
>     .run_commands(
>         "git clone https://github.com/SparkAudio/Spark-TTS.git",
>         "cd Spark-TTS && pip install -q -r requirements.txt",
>         # "git clone https://github.com/pytorch/audio.git",
>         # "cd audio && git checkout c670ad8 && PATH=/usr/local/cuda/bin:$PATH python3 setup.py develop",
>     )
>     .env(
>         {
>             "PYTHONPATH": os.getenv("IMAGE_PYTHONPATH", "/Spark-TTS/"),
>             "CUDA_VISIBLE_DEVICES": os.getenv("CUDA_VISIBLE_DEVICES", "0"),
>             "PYTHONIOENCODING": "utf-8",
>         }
>     )
> )
> 
> 
> HF_MODEL_DIR = "/root/models"
> hf_model_vol = modal.Volume.from_name("models", create_if_missing=True)
> TRT_MODEL_DIR = "/root/trt_models"
> trt_model_vol = modal.Volume.from_name("triton_trtllm_models", create_if_missing=True)
> TRITONSERVER_DIR = "/root/tritonserver"
> tritonserver_vol = modal.Volume.from_name("tritonserver", create_if_missing=True)
> 
> 
> # ⭐️ https://github.com/triton-inference-server/python_backend/blob/main/README.md
> @app.function(
>     gpu=os.getenv("IMAGE_GPU", "L4"),
>     image=tritonserver_image,
>     volumes={
>         HF_MODEL_DIR: hf_model_vol,
>         TRT_MODEL_DIR: trt_model_vol,
>         TRITONSERVER_DIR: tritonserver_vol,
>     },
>     timeout=1200,  # default 300s
>     scaledown_window=1200,
> )
> # @modal.web_server(port=8000, startup_timeout=5 * 60)
> def serve(app_name: str) -> str:
>     import subprocess
> 
>     os.environ["PYTHONPATH"] = f"/usr/local/lib/python3.10/site-packages:{os.environ['PYTHONPATH']}"
> 
>     cmd = f"ldd /opt/tritonserver/bin/tritonserver".split(" ")
>     subprocess.run(cmd, cwd="/")
>     # cmd = f"ldd /opt/tritonserver/backends/tensorrtllm/libtriton_tensorrtllm.so".split(" ")
>     # subprocess.run(cmd, cwd="/")
> 
>     model_repo = os.path.join(TRITONSERVER_DIR, app_name)
> 
>     cmd = f"tritonserver --model-repository {model_repo}"
>     print(cmd)
>     subprocess.run(cmd, env=os.environ, shell=True)
> 
> 
> """
> # run tritonserver
> modal run src/llm/trtllm/bench/tritonserver.py --app-name tts-spark
> """
> 
> 
> @app.local_entrypoint()
> def main(app_name: str = ""):
>     serve.remote(app_name)
> ```
> 
> run cmd `tritonserver --model-repository {model_repo}` have some errors:
> 
> ```
> I0313 17:55:29.117742 80 pb_stub.cc:320]  Failed to initialize Python stub for auto-complete: ImportError: Error importing numpy: you should not try to import numpy from
>         its source directory; please exit the numpy source tree, and relaunch
>         your python interpreter from there.
> 
> At:
>   /usr/local/lib/python3.10/site-packages/numpy/__init__.py(119): <module>
>   <frozen importlib._bootstrap>(488): _call_with_frames_removed
>   <frozen importlib._bootstrap_external>(995): exec_module
>   <frozen importlib._bootstrap>(950): _load_unlocked
>   <frozen importlib._bootstrap>(1333): _find_and_load_unlocked
>   <frozen importlib._bootstrap>(1360): _find_and_load
>   /root/tritonserver/tts-spark/spark_tts/1/model.py(32): <module>
>   <frozen importlib._bootstrap>(488): _call_with_frames_removed
>   <frozen importlib._bootstrap_external>(995): exec_module
>   <frozen importlib._bootstrap>(950): _load_unlocked
>   <frozen importlib._bootstrap>(1334): _find_and_load_unlocked
>   <frozen importlib._bootstrap>(1360): _find_and_load
> 
> E0313 17:55:29.121481 25 model_lifecycle.cc:654] "failed to load 'spark_tts' version 1: Internal: ImportError: Error importing numpy: you should not try to import numpy from\n        its source directory; please exit the numpy source tree, and relaunch\n        your python interpreter from there.\n\nAt:\n  /usr/local/lib/python3.10/site-packages/numpy/__init__.py(119): <module>\n  <frozen importlib._bootstrap>(488): _call_with_frames_removed\n  <frozen importlib._bootstrap_external>(995): exec_module\n  <frozen importlib._bootstrap>(950): _load_unlocked\n  <frozen importlib._bootstrap>(1333): _find_and_load_unlocked\n  <frozen importlib._bootstrap>(1360): _find_and_load\n  /root/tritonserver/tts-spark/spark_tts/1/model.py(32): <module>\n  <frozen importlib._bootstrap>(488): _call_with_frames_removed\n  <frozen importlib._bootstrap_external>(995): exec_module\n  <frozen importlib._bootstrap>(950): _load_unlocked\n  <frozen importlib._bootstrap>(1334): _find_and_load_unlocked\n  <frozen importlib._bootstrap>(1360): _find_and_load\n"
> I0313 17:55:29.121766 25 model_lifecycle.cc:789] "failed to load 'spark_tts'"
> I0313 17:55:29.259034 81 pb_stub.cc:320]  Failed to initialize Python stub for auto-complete: ImportError: Failed to load PyTorch C extensions:
>     It appears that PyTorch has loaded the `torch/_C` folder
>     of the PyTorch repository rather than the C extensions which
>     are expected in the `torch._C` namespace. This can occur when
>     using the `install` workflow. e.g.
>         $ python setup.py install && python -c "import torch"
> 
>     This error can generally be solved using the `develop` workflow
>         $ python setup.py develop && python -c "import torch"  # This should succeed
>     or by running Python from a different directory.
> 
> At:
>   /usr/local/lib/python3.10/site-packages/torch/__init__.py(898): <module>
>   <frozen importlib._bootstrap>(488): _call_with_frames_removed
>   <frozen importlib._bootstrap_external>(995): exec_module
>   <frozen importlib._bootstrap>(950): _load_unlocked
>   <frozen importlib._bootstrap>(1333): _find_and_load_unlocked
>   <frozen importlib._bootstrap>(1360): _find_and_load
>   /root/tritonserver/tts-spark/audio_tokenizer/1/model.py(27): <module>
>   <frozen importlib._bootstrap>(488): _call_with_frames_removed
>   <frozen importlib._bootstrap_external>(995): exec_module
>   <frozen importlib._bootstrap>(950): _load_unlocked
>   <frozen importlib._bootstrap>(1334): _find_and_load_unlocked
>   <frozen importlib._bootstrap>(1360): _find_and_load
> 
> E0313 17:55:29.262042 25 model_lifecycle.cc:654] "failed to load 'audio_tokenizer' version 1: Internal: ImportError: Failed to load PyTorch C extensions:\n    It appears that PyTorch has loaded the `torch/_C` folder\n    of the PyTorch repository rather than the C extensions which\n    are expected in the `torch._C` namespace. This can occur when\n    using the `install` workflow. e.g.\n        $ python setup.py install && python -c \"import torch\"\n\n    This error can generally be solved using the `develop` workflow\n        $ python setup.py develop && python -c \"import torch\"  # This should succeed\n    or by running Python from a different directory.\n\nAt:\n  /usr/local/lib/python3.10/site-packages/torch/__init__.py(898): <module>\n  <frozen importlib._bootstrap>(488): _call_with_frames_removed\n  <frozen importlib._bootstrap_external>(995): exec_module\n  <frozen importlib._bootstrap>(950): _load_unlocked\n  <frozen importlib._bootstrap>(1333): _find_and_load_unlocked\n  <frozen importlib._bootstrap>(1360): _find_and_load\n  /root/tritonserver/tts-spark/audio_tokenizer/1/model.py(27): <module>\n  <frozen importlib._bootstrap>(488): _call_with_frames_removed\n  <frozen importlib._bootstrap_external>(995): exec_module\n  <frozen importlib._bootstrap>(950): _load_unlocked\n  <frozen importlib._bootstrap>(1334): _find_and_load_unlocked\n  <frozen importlib._bootstrap>(1360): _find_and_load\n"
> ```

Sorry, I have no ideas. I believe your error is likely related to environment issues with the modal. Please try reproducing the service with our pre-built Docker image first and ensure that the code works. 

Feel free to raise a GitHub issue if you encounter any errors while using our execution scripts and Docker environment.

weedge commented:
thanks, run local docker it's ok,  but want use modal to deploy, just test.  
maybe  python backend stub different: https://github.com/triton-inference-server/backend/blob/main/docs/python_based_backends.md

weedge commented:
<img width="1433" alt="image" src="https://github.com/user-attachments/assets/1e691189-8081-4662-a2ef-214f66e37903" />
yeap! deploy modal is ok~,  nvcr.io/nvidia/tritonserver:25.02-trtllm-python-py3  use python3.12.3,  modal python3.12 use python3.12.1, need building-custom-python-backend-stub:
https://github.com/triton-inference-server/python_backend?#building-custom-python-backend-stub
